### PR TITLE
TimelockAuthorizer extension to allow for auditability of scheduled tasks onchain

### PR DESCRIPTION
**Pull request by danielmkm**
# Description

First of all, love this contract! This would be a huge step forward for the authorizer.

The one thing I see missing is the ability to provide better insights into scheduled tasks without having to rely on offchain tooling. With some simple extensions to the `ScheduledExecution` struct and a view function, we could make it possible to spin up a pretty robust UI leveraging only RPC calls.

The implementation is presented as a PoC. **IF** you guys think it's worthwhile, will properly add it to the interface with docs and build out tests.

- Extend the ScheduledExecution struct with fields for better auditability
- Add getNumScheduledExecutions to return the length of the array
- Add getScheduledExecutions to allow for fetching part of the array

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [ ] The diff is legible and has no extraneous changes
- [ ] Complex code has been commented, including external interfaces
- [ ] Tests are included for all code paths
- [ ] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

**Comment by danielmkm** [comment]
It is theoretically true that you can leverage `eth_getLogs` to fetch the events and reconcile the data, but in practice this is quite a bit harder than you'd expect. Most RPCs have limitations on the block range you can fetch in a single query, so finding the events for past actions can be challenging. Alongside of that, reconciling the events is error prone and introduces another opportunity for showing incorrect data.

imo, it is important that scheduled actions are as easy as possible to audit, as this is the safeguard that the timelock authorizer is introducing. With the additions in this PR, building the interface below requires 2 `eth_calls`. One call to `getNumScheduledExecutions` and one call to `getScheduledExecutions`.

Screenshot 2023-04-13 at 08 32 30

File: pkg/vault/contracts/authorizer/TimelockAuthorizerManagement.sol [hunk 1]
```diff
@@ -201,6 +201,30 @@ abstract contract TimelockAuthorizerManagement is ITimelockAuthorizer {
         return _scheduledExecutions[scheduledExecutionId];
     }
 
+    function getNumScheduledExecutions() external view returns (uint256) {
```
Comment (nventuro) [inline pkg/vault/contracts/authorizer/TimelockAuthorizerManagement.sol:204]:
```suggestion
    function getScheduledExecutionsCount() external view returns (uint256) {
```

File: pkg/vault/contracts/authorizer/TimelockAuthorizerManagement.sol [hunk 2]
```diff
@@ -201,6 +201,30 @@ abstract contract TimelockAuthorizerManagement is ITimelockAuthorizer {
...
+
+    function getScheduledExecutions(uint256 offset, uint256 limit, bool reverseOrder)
+        external
+        view
+        returns (ITimelockAuthorizer.ScheduledExecution[] memory items)
+    {
+        require(offset < _scheduledExecutions.length, "INVALID_OFFSET");
+      
+        uint256 rest = _scheduledExecutions.length - offset;
+        uint256 size = rest > limit ? limit : rest;
```
Comment (nventuro) [inline pkg/vault/contracts/authorizer/TimelockAuthorizerManagement.sol:216]:
```suggestion
        uint256 size = Math.min(rest, limit);
```

I find this makes reasoning about `size` easier (i.e. that `size` is never greater than `rest`, and that `limit` may reduce `rest`).

File: pkg/vault/contracts/authorizer/TimelockAuthorizerManagement.sol [hunk 3]
```diff
@@ -286,6 +310,9 @@ abstract contract TimelockAuthorizerManagement is ITimelockAuthorizer {
         }
 
         scheduledExecution.executed = true;
+        scheduledExecution.executedBy = msg.sender;
+        // solhint-disable-next-line not-rely-on-time
```
Comment (nventuro) [inline pkg/vault/contracts/authorizer/TimelockAuthorizerManagement.sol:314]: I'd move this to the top of the file, disable it for the entire file, and add a short comment mentioning that timelocks are time based, and expected to be in the order of days, so any time changes or manipulation in the order of seconds or minutes are irrelevant.

**Review by nventuro — COMMENTED** [review]
Been thinking about this and I like it. Other data structures are too complex to iterate over, but this one specifically does make a lot of sense I think. Thanks for the suggestion!

@ylv-io curious to hear your thoughts on this matter

**Review by ylv-io — APPROVED** [review]
It is a great value addition. In fact, we didn't even consider this, which I now feel ashamed of. Given that TimelockAuthorizer is not a user-facing contract, we do not care that much about gas costs, but the value added from being able to fetch data on-chain is significant.

I would like to second @danielmkm's experience regarding fetching eth_logs. It is neither reliable nor easy to do on a frontend.

**Comment by nventuro** [comment]
Superceded by https://github.com/balancer/balancer-v2-monorepo/pull/2485, thanks a lot @danielmkm!

**Pull request closed without merging**